Add bin/sync

schrockwell opened the pull request:
## Overview
This PR adds the `bin/sync` executable to ease the creation of new keys. (#71)

The flow is simple: add translations to an `locales/*/en.yml` file, then run `bin/sync` to copy it to all the other locales. Raw values are generated, too.

A follow-up PR will merge in the normalized translation files.

schrockwell commented on bin/sync:
```diff
@@ -0,0 +1,9 @@
+#! /usr/bin/env ruby
+# frozen_string_literal: true
+
+# rubocop:disable Packaging/RequireRelativeHardcodingLib
+require_relative "../lib/trmnl/i18n/editor"
+# rubocop:enable Packaging/RequireRelativeHardcodingLib
```
@bkuhlmann I couldn't find a way around this. The autocorrected solution did not work.

schrockwell commented on lib/trmnl/i18n/editor.rb:
```diff
@@ -0,0 +1,64 @@
+module TRMNL
+  module I18n
+    # Easy manipulation of locale files
+    class Editor
```
Yeah, that is a better fit. When I started this I was going to have multiple methods for adding, removing, etc., but ended up on a single sync.

bkuhlmann replied:
Ah, that makes more sense now. Ha, yep, happens to me a lot too so I end up rebasing so the final implementation uses the new terminology and you never knew I had a different name. :wink:

schrockwell commented on lib/trmnl/i18n/editor.rb:
```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+require "active_support/core_ext/hash/deep_merge"
```
active_support is also already in the gemfile?

bkuhlmann replied:
True and a fair point. I'm not a fan of Active Support since it monkey patches the entire object space. With refinements, you have lexical scope, which means you only alter the file and/or class/module you are in and nothing else.

...but if you want to stick with Active Support, that works too. I guess we can't avoid it since this is a lightweight Rails Engine.

schrockwell commented on lib/trmnl/i18n/raw_locale.rb:
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+module TRMNL
```
I split it out like this to appease one of the Reek rules. I guess I could make these static methods on the class instead?

bkuhlmann replied:
So static class methods would work but they are a different code smell because you'd only be moving the problem out of a module and into a class (those methods are still static methods just in a different location now).

Whenever, I hit that situation, myself, it's always a realization that I have an object that hasn't been given a name yet (channeling my inner Sandi Metz here). Once I give the object a name, that means I can inject it into the class that needs to use it (i.e. Dependency Injection). This makes testing a breeze and also means you an inject spies for testing purposes to swap out behavior if need be. Lots of power when you wield object composition like this.

Anyway, give it a shot and if composition helps and feels better. Happy to pair up if that'd help too!

schrockwell replied:
I came up with a solution that has some nice parallels - a base `Generator` for handling any type, and subclasseses `HashGenerator` and `ArrayGenerator` for those specific types. Added some specs as well.

bkuhlmann replied:
Thanks. This is looking better and it's nice to have specs added (although there is missing spec for the `Syncer` itself).

Interestingly, the specs show how this is still hard to test because there are so many contexts used to setup the object. Also, there are no subjects only lets (💡 this is also known as the the [Misused Lets](https://alchemists.io/articles/rspec_antipatterns#_misused_lets) antipattern).

If you passed in the values that *change* to each `#call` method, then you'd have a more resuable object and easier to test too because the data that is changing happens at runtime, not initialization. In other words what you want is to inject your dependencies via your constructor and them supply the `value` and `path` to each `#call` method if that make sense.

schrockwell commented on .reek.yml:
```diff
@@ -5,3 +5,5 @@ exclude_paths:
 detectors:
   LongParameterList:
     enabled: false
+  TooManyStatements:
+    enabled: false
```
To me, that seems like a very overcomplicated solution. Splitting out all those responsibilities without any idea about how they might be reused commits the same cardinal sin as premature optimization. 

I am really not used to having all these arbitrary restrictions.

bkuhlmann replied:
> seems like a very overcomplicated solution

Only if you've never been taught to break objects down into compossible/functional components. This is why I spend so much time writing and teaching others about good OO/FP design. I'm not the only one to architect this way. [Hanami](https://hanamirb.org) has the same philosophy, [Elm](http://elm-lang.org) too, and [POODR](https://www.poodr.com) is written that way to name a few.

> I am really not used to having all these arbitrary restrictions.

Don't worry, it gets much easier in time. You're learning to build up a new mental muscle that'll make you stronger. Always a good thing. :tada:

schrockwell commented on lib/trmnl/i18n/editor.rb:
```diff
@@ -0,0 +1,64 @@
+
+require_relative "raw_locale"
+
+module TRMNL
```
Sure, I will add a spec, but the comments were required to satisfy one of the linters.

bkuhlmann replied:
Thanks. So, with comments, you really only want to add them if only to clarify. Otherwise, they add additional maintenance that, if not upkept, will lead people astray. If you focus on keeping the code readable, then the comments become no longer necessary.

bkuhlmann reviewed (commented):
How about we fold this in and I can refactor a bit further to show how use composition instead of inheritance? Maybe that would be more useful and would allow me to illustrate what I'm talking about more through code than through text?

By the way, if you run this syncer locally did you notice that all of the YAML files end up with single quotes? Is that we want want?

schrockwell commented:
> How about we fold this in and I can refactor a bit further to show how use composition instead of inheritance? Maybe that would be more useful and would allow me to illustrate what I'm talking about more through code than through text?

Sure, let me finish up this latest iteration I have in-flight, and then you can take it over.

> By the way, if you run this syncer locally, did you notice that all of the YAML files end up with single quotes? Is that we want?

I don't have a preference, as long as it's consistent going forward. The built-in YAML formatter does not appear to have an option for this, I'm not keen about adding a dependency to fix it, and I don't feel confident in building a custom solution to handle it.

schrockwell commented:
@bkuhlmann You may take over this PR now. It's fully tested and working. I have incorporated the suggestions to the best of my ability. 

Here are the changes I made since the initial review (I can't share a diff because the changes were squashed):

1. Renamed `Editor` to `Syncer`
2. Encapsulated YAML access into a `Repo` class, decoupling the filesystem from the `Syncer`
3. Encapsulated generation of the `raw` locale into three `Generator` classes for handling the various nested types (hashes, arrays, etc)
4. Added specs
5. Removed linter exceptions

bkuhlmann commented:
> The built-in YAML formatter does not appear to have an option for this, I'm not keen about adding a dependency to fix it, and I don't feel confident in building a custom solution to handle it.

Yeah, interesting. There might be a setting for but don't remember off the top of my head but will look.

> You may take over this PR now. It's fully tested and working.

Awesome, thanks for doing all this! I'll polish this up a bit further and get this folded. I'm not sure if I can get to this tonight (might be tomorrow).

bkuhlmann commented:
:information_source: Ugh, accidentally ran the wrong command. The branch is restored again, sorry for the noise. 🤦

bkuhlmann commented:
:information_source: Need to take another pass at this in the morning to finish this up but did the following:

- Moved all objects to the `Synchronization` to properly namespace all objects as one cohesive unit.
- Removed the use of `File` in favor of `Pathname` since it has a easier to use Object API.
- Updated all specs to use `SPEC_ROOT` and use proper subjects.
- Used `private` scope instead of `protected` scope as you only want to use `protected` scope when a subclass needs to leverage a superclass' method.

bkuhlmann commented:
:information_source: OK, I simplified the original generator logic by using the new `ValueReducer` function. There's a couple of reasons for this change:

- Using a single function to deal with Array/Hash recursion is easier to reason about and test.
- By using a function, we no longer need to deal with generator subclasses which introduced circular dependencies and made the code to tangled upon itself.

schrockwell commented on lib/trmnl/i18n/synchronization/value_reducer.rb:
```diff
@@ -0,0 +1,23 @@
+  module I18n
+    module Synchronization
+      # Reduces a nested structure into key/value pairs with flattened values.
+      ValueReducer = lambda do |initial_value, path = ""|
```
This function is very similar to what I had originally written. Doesn't it break the 5-statements-per-method rule?
